test/device/add_inlink_test.cpp

```diff
@@ -0,0 +1,51 @@
+#include <gtest/gtest.h>
+
+#include "utils.hpp"
+
+namespace test {
+
+class AddInlink : public testing::Test {
+public:
+    void TearDown() override {};
+    void SetUp() override {};
+};
+
+TEST_F(AddInlink, LinkIsPresent) {
+    auto source = std::make_shared<sim::RoutingModule>(sim::RoutingModule());
+    auto dest = std::make_shared<sim::RoutingModule>(sim::RoutingModule());
+    auto link = std::make_shared<TestLink>(TestLink(source, dest));
+
+    EXPECT_EQ(dest->next_inlink(), nullptr);
+    dest->add_inlink(link);
+    EXPECT_EQ(dest->next_inlink(), link);
+}
+
+TEST_F(AddInlink, SameLinkTwice) {
+    int NUMBER_OF_ITERATIONS = 6;
+    int FIRST_LINK_VALUE = 1;
+    int SECOND_LINK_VALUE = 10;
+    int EXPECTED_RESULT =
+        (FIRST_LINK_VALUE + SECOND_LINK_VALUE) * (NUMBER_OF_ITERATIONS / 2);
+
+    auto source = std::make_shared<sim::RoutingModule>(sim::RoutingModule());
+    auto doubled_source =
+        std::make_shared<sim::RoutingModule>(sim::RoutingModule());
+    auto dest = std::make_shared<sim::RoutingModule>(sim::RoutingModule());
+
+    auto link = std::make_shared<TestLink>(
+        TestLink(source, dest, sim::Packet(sim::DATA, 1)));
+    auto doubled_link = std::make_shared<TestLink>(
+        TestLink(source, dest, sim::Packet(sim::DATA, 10)));
+
+    dest->add_inlink(doubled_link);
+    dest->add_inlink(link);
+    dest->add_inlink(doubled_link);
+
+    int sum = 0;
+    for (int i = 0; i < NUMBER_OF_ITERATIONS; i++) {
+        sum += dest->next_inlink()->get_packet()->size;
+    }
+    EXPECT_EQ(sum, EXPECTED_RESULT);
+}
```

Comment:
In my oppinion this test should be simply. E.g:
```suggestion
std::vecotr<std::shared_ptr<sim::Link> > get_inlinks(std::shared_ptr<sim::RoutingModule> device) {
    std::vecotr<std::shared_ptr<sim::Link> > inlinks;
    auto next_inlink = device->next_inlink();
    do {
       inlinks.push_back(next_inlink);
       next_inlink = device->next_inlink();
    } while (next_inlink != inlinks[0]);
    return inlinks;
}

TEST_F(AddInlink, SameLinkTwice) {
    auto source = std::make_shared<sim::RoutingModule>();
    auto destination = std::make_shared<sim::RoutingModule>();

    auto link = std::make_shared<TestLink>(source, dest, sim::Packet(sim::DATA, 0));
    
    dest->add_inlink(link);
    dest->add_inlink(link);
    
    EXPECT_EQ(get_inlinks(dest).size(), 1);
}
```
I think the simpliest test like this one is necessary. Also more complex test might be userful (e.g. add some count of links several times in random order)

Reply:
I was trying to do something like this, but it does not work here
If we have real inlinks storage like [link, link] the function get_inlinks will return [link] and the test will pass
So we at least need to add one more link to the test but such test will still be hardly depended on the exact order of links in real storage (like if for some reason links will be stored like [link, other_link, link] the test will pass, but it's not what we want to have)

Reply:
I designed this test only for checking that adding link twice have no affect, not to check the order of adding links